### PR TITLE
Making Verbnet feature more granular

### PR DESCRIPTION
**Pull request by sbhaktha**
@mhrmm : Changes as discussed. Some improvement in scores.
Note that based on the table at the top of this page: http://verbs.colorado.edu/~mpalmer/projects/verbnet.html#frames, you had suggested using the Frame Name and the Frame Syntax. For the e.g. in that table (`hit-18.1`), this is what the frame xml snippet looks like:

```
        <FRAME>
            <DESCRIPTION descriptionNumber="0.2" primary="NP V NP" secondary="Basic Transitive" xtag="0.2"/>
            <EXAMPLES>
                <EXAMPLE>Paula hit the ball.</EXAMPLE>
            </EXAMPLES>
            <SYNTAX>
                <NP value="Agent">
                    <SYNRESTRS/>
                </NP>
                <VERB/>
                <NP value="Patient">
                    <SYNRESTRS/>
                </NP>
            </SYNTAX>
...
```

`Basic Transitive` is the secondary name, and `NP V NP` is the primary frame name, which represents the syntax `Agent V Patient`. So I have used the primary frame names and the secondary frame names as the features. Since a class can have multiple frames, there can be multiple values (a set) coming from the same class for a given verb.

The `featuretaggers.config` (rename of the earlier `polyparser.config`) can be used to pass any config info for any other features, for e.g., the Brown clusters, which is not in the datastore yet.

Scores

Baseline

```
UAS: 9084 / 10427 = 87.12%
UAS (no punc): 8019 / 9130 = 87.83%
LAS: 8663 / 10427 = 83.08%
LAS (no punc): 7598 / 9130 = 83.22%
Path Accuracy (labeled, full): 7371 / 10427 = 70.69%
Path Accuracy (unlabeled, full): 7903 / 10427 = 75.79%
Path Accuracy (labeled, ignorePunc): 6347 / 9130 = 69.52%
Path Accuracy (unlabeled, ignorePunc): 6875 / 9130 = 75.30%
Parsed 1000 sentences in 16.3 seconds, an average of 61.4 sentences per second.
```

With Verbnet

```
UAS: 9142 / 10427 = 87.68%
UAS (no punc): 8059 / 9130 = 88.27%
LAS: 8741 / 10427 = 83.83%
LAS (no punc): 7659 / 9130 = 83.89%
Path Accuracy (labeled, full): 7515 / 10427 = 72.07%
Path Accuracy (unlabeled, full): 8007 / 10427 = 76.79%
Path Accuracy (labeled, ignorePunc): 6465 / 9130 = 70.81%
Path Accuracy (unlabeled, ignorePunc): 6954 / 9130 = 76.17%
Parsed 1000 sentences in 19.4 seconds, an average of 51.4 sentences per second.
```

With Brown Clusters

```
UAS: 9154 / 10427 = 87.79%
UAS (no punc): 8084 / 9130 = 88.54%
LAS: 8780 / 10427 = 84.20%
LAS (no punc): 7710 / 9130 = 84.45%
Path Accuracy (labeled, full): 7505 / 10427 = 71.98%
Path Accuracy (unlabeled, full): 8027 / 10427 = 76.98%
Path Accuracy (labeled, ignorePunc): 6462 / 9130 = 70.78%
Path Accuracy (unlabeled, ignorePunc): 6980 / 9130 = 76.45%
Parsed 1000 sentences in 20.0 seconds, an average of 49.9 sentences per second.
```

With Verbnet + Brown Clusters

```
UAS: 9215 / 10427 = 88.38%
UAS (no punc): 8114 / 9130 = 88.87%
LAS: 8835 / 10427 = 84.73%
LAS (no punc): 7736 / 9130 = 84.73%
Path Accuracy (labeled, full): 7570 / 10427 = 72.60%
Path Accuracy (unlabeled, full): 8090 / 10427 = 77.59%
Path Accuracy (labeled, ignorePunc): 6510 / 9130 = 71.30%
Path Accuracy (unlabeled, ignorePunc): 7026 / 9130 = 76.96%
Parsed 1000 sentences in 24.4 seconds, an average of 41.0 sentences per second.
```
Side note: Training takes a lot longer with the Verbnet features, as recorded in the final timing line.

### CODE REVIEW

**Comment by mhrmm** [comment]
Cool! One of the unit tests appears to be failing though...

**Comment by sbhaktha** [comment]
@mhrmm : Check it out now. Test passes. Embarrassing bug fixed in this commit: https://github.com/sbhaktha/nlpstack/commit/c4f32c1a09fc2deeccd7103700c282706d763703 .

Scores:

Primary Frames Only

```
[info] UAS: 9138 / 10427 = 87.64%
[info] UAS (no punc): 8061 / 9130 = 88.29%
[info] LAS: 8737 / 10427 = 83.79%
[info] LAS (no punc): 7660 / 9130 = 83.90%
[info] Path Accuracy (labeled, full): 7465 / 10427 = 71.59%
[info] Path Accuracy (unlabeled, full): 7961 / 10427 = 76.35%
[info] Path Accuracy (labeled, ignorePunc): 6430 / 9130 = 70.43%
[info] Path Accuracy (unlabeled, ignorePunc): 6921 / 9130 = 75.81%
[info] Parsed 1000 sentences in 20.4 seconds, an average of 49.0 sentences per second.
```

Secondary Frames Only

```
[info] UAS: 9142 / 10427 = 87.68%
[info] UAS (no punc): 8059 / 9130 = 88.27%
[info] LAS: 8741 / 10427 = 83.83%
[info] LAS (no punc): 7659 / 9130 = 83.89%
[info] Path Accuracy (labeled, full): 7515 / 10427 = 72.07%
[info] Path Accuracy (unlabeled, full): 8007 / 10427 = 76.79%
[info] Path Accuracy (labeled, ignorePunc): 6465 / 9130 = 70.81%
[info] Path Accuracy (unlabeled, ignorePunc): 6954 / 9130 = 76.17%
[info] Parsed 1000 sentences in 19.4 seconds, an average of 51.4 sentences per second.
```

Primary + Secondary

```
[info] UAS: 9119 / 10427 = 87.46%
[info] UAS (no punc): 8044 / 9130 = 88.11%
[info] LAS: 8708 / 10427 = 83.51%
[info] LAS (no punc): 7633 / 9130 = 83.60%
[info] Path Accuracy (labeled, full): 7443 / 10427 = 71.38%
[info] Path Accuracy (unlabeled, full): 7944 / 10427 = 76.19%
[info] Path Accuracy (labeled, ignorePunc): 6404 / 9130 = 70.14%
[info] Path Accuracy (unlabeled, ignorePunc): 6903 / 9130 = 75.61%
[info] Parsed 1000 sentences in 21.9 seconds, an average of 45.7 sentences per second.
```

If `verbnet` feature requested (per training cmd line), by default both features are included currently. But `verbnet` is off by default anyway. If you think I should change it so only Secondary frame features are included (based on max scores above) lmk.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/Training.scala [hunk 1]
```diff
@@ -57,29 +63,47 @@ object Training {
...
+    } {
+      val verbnetPath: java.nio.file.Path = Datastore.directoryPath(
+        groupName,
+        artifactName,
+        version
+      )
+      val verbnet = new Verbnet(verbnetPath)
+      taggers :+= VerbnetTagger(verbnet)
+    }
+
```
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/Training.scala]: You can get away without using a var:

```
    val maybeVerbnetTagger: Option[VerbnetTagger] = for {
      taggersConfig <- taggersConfigOption
      verbnetConfig <- taggersConfig.get[Config]("verbnet")
      groupName <- verbnetConfig.get[String]("group")
      artifactName <- verbnetConfig.get[String]("name")
      version <- verbnetConfig.get[Int]("version")
    } yield {
      val verbnetPath: java.nio.file.Path = Datastore.directoryPath(
        groupName,
        artifactName,
        version
      )
      VerbnetTagger(new Verbnet(verbnetPath))
    }
    val taggers: Seq[SentenceTransform] =
      Seq(FactorieSentenceTagger, LexicalPropertiesTagger,
        BrownClustersTagger(clusters), maybeVerbnetTagger).flatten
```

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    val secondaryFrameNames = new scala.collection.mutable.HashSet[Symbol]
+    for (verbFrame <- verbFrames) {
+      val secondaryType = verbFrame.getSecondaryType
+      if (secondaryType != null) {
+        secondaryFrameNames += Symbol(secondaryType.getID)
+      }
+    }
+    secondaryFrameNames.toSet
+  }
+}
```
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala]: Replacing every sequence of whitespace in the frame names with something (e.g. a hyphen) will make the feature names more readable during debugging... e.g. right now they look like:

s1.verbnetSecondaryFrames.Infinitival Copular Clause

**Comment by mhrmm** [comment]
Looks very good. A couple of comments. Also, generating a histogram of all primary and secondary frames (per our discussion yesterday) may allow us to find more patterns we can exploit.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/Training.scala (hunk 1 above)
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/Training.scala]: I looked this up today and read somewhere that `Option` is like a collection, so just doing a `++ maybeVerbnetTagger` works.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala (hunk 2 above)
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala]: done

**Comment by sbhaktha** [comment]
@mhrmm : Check out latest commit, addresses above points. 

Also dumped out some Verbnet data and stats today. Here are some histograms, though I am not sure this is what you had in mind. I can iterate with you on it, but here are some:

https://drive.google.com/drive/#folders/0B971Sd5f2y4HfmkyaHdnZy01RWtBR3dISlo3bmJ1bE1YSndSYUZ2SVdhYmdiYjRWQUJNams

`primaryFrameVerbStats, secondaryFrameVerbStats : `
x-axis : verb count bins
y-axis : no. of primary/secondary frames respectively
— can be read as, for e.g., “most primary frames have 1-5 different verbs associated with them”.


`verbPrimaryFrameStats, verbSecondaryFrameStats:`
x-axis : primary/secondary frame count bins respectively
y-axis : no. of verbs
— can be read as, for e.g., “most verbs occupy 3-4 different primary frames”.

**Comment by sbhaktha** [comment]
I also have plots of actual frames to no. of verbs, and the other way round to get an idea of the distribution. I'll put those up tonight as well.

**Comment by mhrmm** [comment]
Could you give me permission to view the above Google Drive link? Thanks!

**Comment by sbhaktha** [comment]
Can you try now? https://drive.google.com/a/allenai.org/folderview?id=0B971Sd5f2y4HfmkyaHdnZy01RWtBR3dISlo3bmJ1bE1YSndSYUZ2SVdhYmdiYjRWQUJNams&usp=sharing

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    verbClasses.map(x => Symbol(x.getID))
+  }
+
+  /* Returns the set of all frames within all classes for a given verb. 
+   */
+  def getVerbnetFrames(verb: String): Set[IFrame] = {
+    // Get all classes
+    val verbClasses = getVerbnetClasses(verb)
+    // Create set of verb frames to populate
+    val verbFrames = new scala.collection.mutable.HashSet[IFrame]
```
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala]: Why use a mutable HashSet here? I think you can replace the contents of this function with the following one-liner:

```
getVerbnetClasses(verb) flatMap { verbClass => verbClass.getFrames() }
```

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    verbFrames.map(x => Symbol(x.getPrimaryType.getID.replaceAll("""\s+""", "-")))
+  }
+
+  /* Returns the set of secondary names for all frames within all classes 
+   * for a given verb. 
+   */
+  def getVerbnetFrameSecondaryNames(verb: String): Set[Symbol] = {
+    // Get frames
+    val verbFrames = getVerbnetFrames(verb)
+    val secondaryFrameNames = new scala.collection.mutable.HashSet[Symbol]
```
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala]: Same comment as above applies here as well, I believe.

Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala]: Well there's the null check in the middle, but I converted the whole thing into calls to `map` and `filter`.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala (hunk 3 above)
Comment (sbhaktha) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/Verbnet.scala]: right. done.

**Comment by sbhaktha** [comment]
@mhrmm Using only primary frames. Removed role suffixes such as `.Theme` from syntactic constituents.
Fixes based on your two above comments.

**Comment by mhrmm** [comment]
LGTM!

**Pull request merged**